spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * MongoDbMetadataStore implementation of {@link ConcurrentMetadataStore}.
+ * Use this {@link org.springframework.integration.metadata.MetadataStore} to
+ * achieve meta-data persistence shared across application instances and
+ * restarts.
+ *
+ * @author SenthilArumugam SamirajPanneerSelvam
+ * @since 4.0
+ *
+ */
+public class MongoDbMetadataStore implements ConcurrentMetadataStore {
+
+	private MongoTemplate template;
+	private final static String DEFAULT_COLLECTION_NAME = "metadatastore";
+	private String collectionName;
+
+	/**
+	 * 
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
+	 * @param collectionName
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		this.template = template;
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
+		this(template, null);
+	}
+
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore fileInfo = new Datastore(key, value);
+		if (!template.exists(query, collectionName)) {
+			template.save(fileInfo, collectionName);
+		}
+	}
+
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore result = template.findOne(query, Datastore.class, collectionName);
+		if(result != null){
+			return result.value;	
+		}
+		return null;
+		
+	}
+
+	@Override
+	public String remove(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore result = template.findAndRemove(query, Datastore.class,collectionName);
+		return result==null?null:result.value;
+	}
+
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+
+		String result = value;
+		Query query = new Query(Criteria.where("key").is(key));
+		if (template.exists(query, collectionName)) {
+			result = null;
+		} else {
+			template.save(new Datastore(key, value), collectionName);
+		}
+		return result;
+	}
+
+	@Override
+	public boolean replace(String key, String oldValue, String newValue) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(oldValue, "'oldValue' must not be null.");
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		
+		Query query = new Query(Criteria.where("key").is(key));
```

Comment:
You have missed `oldValue` in the criteria.
See the `replace` contract in the `ConcurrentMap.replace`